provider/consul/consul_catalog.go

```diff
@@ -93,38 +100,100 @@ func (p *CatalogProvider) watchServices(stopCh <-chan struct{}) <-chan map[strin
 			default:
 			}
 
-			data, catalogMeta, err := catalog.Services(catalogOptions)
-			if err != nil {
-				log.WithError(err).Error("Failed to list services")
-				return
-			}
-
 			// Listening to changes that leads to `passing` state or degrades from it.
 			// The call is used just as a trigger for further actions
 			// (intentionally there is no interest in the received data).
-			_, healthMeta, err := health.State("passing", healthOptions)
+			_, meta, err := health.State("passing", options)
 			if err != nil {
 				log.WithError(err).Error("Failed to retrieve health checks")
 				return
 			}
 
 			// If LastIndex didn't change then it means `Get` returned
 			// because of the WaitTime and the key didn't changed.
-			sameServiceAmount := catalogOptions.WaitIndex == catalogMeta.LastIndex
-			sameServiceHealth := lastHealthIndex == healthMeta.LastIndex
-			if sameServiceAmount && sameServiceHealth {
+			if options.WaitIndex == meta.LastIndex {
 				continue
 			}
-			catalogOptions.WaitIndex = catalogMeta.LastIndex
-			lastHealthIndex = healthMeta.LastIndex
+
+			options.WaitIndex = meta.LastIndex
+
+			// The response should be unified with watchCatalogServices
+			data, _, err := catalog.Services(&api.QueryOptions{})
+			if err != nil {
+				log.Errorf("Failed to list services: %s", err)
+				return
+			}
 
 			if data != nil {
-				watchCh <- data
+				// A critical note is that the return of a blocking request is no guarantee of a change.
+				// It is possible that there was an idempotent write that does not affect the result of the query.
+				// Thus it is required to do extra check for changes...
+				addedKeys, removedKeys := getChangedKeys(data, flashback)
+
+				if len(addedKeys) > 0 {
+					log.WithField("DiscoveredServices", addedKeys).Debug("Health State change detected.")
+					watchCh <- data
+					flashback = data
+				}
+
+				if len(removedKeys) > 0 {
+					log.WithField("MissingServices", removedKeys).Debug("Health State change detected.")
+					watchCh <- data
+					flashback = data
+				}
 			}
 		}
 	})
+}
+
+func (p *CatalogProvider) watchCatalogServices(stopCh <-chan struct{}, watchCh chan<- map[string][]string) {
+	catalog := p.client.Catalog()
+
+	safe.Go(func() {
+		// variable to hold previous state
+		var flashback map[string][]string
+
+		options := &api.QueryOptions{WaitTime: DefaultWatchWaitTime}
 
-	return watchCh
+		for {
+			select {
+			case <-stopCh:
+				return
+			default:
+			}
+
+			data, meta, err := catalog.Services(options)
+			if err != nil {
+				log.Errorf("Failed to list services: %s", err)
+				return
+			}
+
+			if options.WaitIndex == meta.LastIndex {
+				continue
+			}
+
+			options.WaitIndex = meta.LastIndex
+
+			if data != nil {
```

Comment:
I don't think anything should be placed in the 'else' block. The explanation is given in the earlier comment.